<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="This function implements a common workflow to fit linear models
and compute moderated moderated t-statistic by empirical Bayes
moderation of the standard errors towards a global value.
If the input data was generated with microarray technology, the following
steps are executed:

linear model fitting for each feature: lmFit
compute coefficients and standard errors for a given set of contrasts: contrasts.fit
compute moderated t-statistics: eBayes
summarise the linear model fit: topTable


If the input data was generated with RNA-seq technology, the following
steps are executed:

(optional) transform count data for linear modelling: voom
linear model fitting for each feature: lmFit
compute coefficients and standard errors for a given set of contrasts: contrasts.fit
compute moderated t-statistics: eBayes
summarise the linear model fit: topTable


For complete details of each step see the manual pages of the respective
functions."><title>Empirical Bayes Moderated t-Statistic — rowModeratedT • featscreen</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Empirical Bayes Moderated t-Statistic — rowModeratedT"><meta property="og:description" content="This function implements a common workflow to fit linear models
and compute moderated moderated t-statistic by empirical Bayes
moderation of the standard errors towards a global value.
If the input data was generated with microarray technology, the following
steps are executed:

linear model fitting for each feature: lmFit
compute coefficients and standard errors for a given set of contrasts: contrasts.fit
compute moderated t-statistics: eBayes
summarise the linear model fit: topTable


If the input data was generated with RNA-seq technology, the following
steps are executed:

(optional) transform count data for linear modelling: voom
linear model fitting for each feature: lmFit
compute coefficients and standard errors for a given set of contrasts: contrasts.fit
compute moderated t-statistics: eBayes
summarise the linear model fit: topTable


For complete details of each step see the manual pages of the respective
functions."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">featscreen</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/featscreen.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Screening</h6>
    <a class="dropdown-item" href="../articles/supervised.html">Supervised screening methods</a>
    <a class="dropdown-item" href="../articles/unsupervised.html">Unupervised screening methods</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/alebarberis/featscreen/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Empirical Bayes Moderated t-Statistic</h1>
      <small class="dont-index">Source: <a href="https://github.com/alebarberis/featscreen/blob/HEAD/R/4-biostats-test-functions.R" class="external-link"><code>R/4-biostats-test-functions.R</code></a></small>
      <div class="d-none name"><code>rowModeratedT.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function implements a common workflow to fit linear models
and compute moderated moderated t-statistic by empirical Bayes
moderation of the standard errors towards a global value.</p>
<p>If the input data was generated with microarray technology, the following
steps are executed:</p>
<ol><li><p>linear model fitting for each feature: <code><a href="https://rdrr.io/pkg/limma/man/lmFit.html" class="external-link">lmFit</a></code></p></li>
<li><p>compute coefficients and standard errors for a given set of contrasts: <code><a href="https://rdrr.io/pkg/limma/man/contrasts.fit.html" class="external-link">contrasts.fit</a></code></p></li>
<li><p>compute moderated t-statistics: <code><a href="https://rdrr.io/pkg/limma/man/ebayes.html" class="external-link">eBayes</a></code></p></li>
<li><p>summarise the linear model fit: <code><a href="https://rdrr.io/pkg/limma/man/toptable.html" class="external-link">topTable</a></code></p></li>
</ol><p>If the input data was generated with RNA-seq technology, the following
steps are executed:</p>
<ol><li><p>(optional) transform count data for linear modelling: <code><a href="https://rdrr.io/pkg/limma/man/voom.html" class="external-link">voom</a></code></p></li>
<li><p>linear model fitting for each feature: <code><a href="https://rdrr.io/pkg/limma/man/lmFit.html" class="external-link">lmFit</a></code></p></li>
<li><p>compute coefficients and standard errors for a given set of contrasts: <code><a href="https://rdrr.io/pkg/limma/man/contrasts.fit.html" class="external-link">contrasts.fit</a></code></p></li>
<li><p>compute moderated t-statistics: <code><a href="https://rdrr.io/pkg/limma/man/ebayes.html" class="external-link">eBayes</a></code></p></li>
<li><p>summarise the linear model fit: <code><a href="https://rdrr.io/pkg/limma/man/toptable.html" class="external-link">topTable</a></code></p></li>
</ol><p>For complete details of each step see the manual pages of the respective
functions.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">rowModeratedT</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  y <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  observations <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  technology <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"array"</span>, <span class="st">"seq"</span><span class="op">)</span>,</span>
<span>  is.logged <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  mean.variance <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ebayes"</span>, <span class="st">"weights"</span><span class="op">)</span>,</span>
<span>  span <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ls"</span>, <span class="st">"robust"</span><span class="op">)</span>,</span>
<span>  design <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  weights <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ndups <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  spacing <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  block <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  correlation <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  contrasts <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  proportion <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  stdev.coef.lim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.1</span>, <span class="fl">4</span><span class="op">)</span>,</span>
<span>  robust <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  winsor.tail.p <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.05</span>, <span class="fl">0.1</span><span class="op">)</span>,</span>
<span>  coef <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  adjust.method <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>  logger <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>x</dt>
<dd><p>a matrix-like data object with rows corresponding to genes and
columns to observations.</p></dd>


<dt>y</dt>
<dd><p>a vector, factor or matrix. It is used to create a design matrix if
not explicitly provided via the <code>design</code> argument.</p></dd>


<dt>observations</dt>
<dd><p>(optional) integer vector, the indices of observations to
keep.</p></dd>


<dt>technology</dt>
<dd><p>character string, the technology used to generate the data.
Available options are:</p><dl><dt>array</dt>
<dd><p>data generated with microarray technology</p></dd>

<dt>seq</dt>
<dd><p>data generated with RNA-seq technology</p></dd>


</dl></dd>


<dt>is.logged</dt>
<dd><p>logical, whether the original data is already logged. If
<code>is.logged = FALSE</code>, the data is internally transformed.</p></dd>


<dt>mean.variance</dt>
<dd><p>character string indicating whether the mean-variance
relationship should be modeled with precision weights (<code>mean.variance =
"weights"</code>) or with an empirical Bayes prior trend (<code>mean.variance =
"ebayes"</code>).</p></dd>


<dt>span</dt>
<dd><p>width of the smoothing window used for the lowess mean-variance trend.
    Expressed as a proportion between 0 and 1.</p></dd>


<dt>method</dt>
<dd><p>fitting method; <code>"ls"</code> for least squares or <code>"robust"</code> for robust regression</p></dd>


<dt>design</dt>
<dd><p>the design matrix of the microarray experiment, with rows corresponding to samples and columns to coefficients to be estimated.
  Defaults to <code>object$design</code> if that is non-NULL, otherwise to the unit vector meaning that all samples will be treated as replicates of a single treatment group.</p></dd>


<dt>weights</dt>
<dd><p>non-negative precision weights.  Can be a numeric matrix of individual weights of same size as the object expression matrix, or a numeric vector of array weights with length equal to <code>ncol</code> of the expression matrix, or a numeric vector of gene weights with length equal to <code>nrow</code> of the expression matrix.</p></dd>


<dt>ndups</dt>
<dd><p>positive integer giving the number of times each distinct probe is printed on each array.</p></dd>


<dt>spacing</dt>
<dd><p>positive integer giving the spacing between duplicate occurrences of the same probe, <code>spacing=1</code> for consecutive rows.</p></dd>


<dt>block</dt>
<dd><p>vector or factor specifying a blocking variable on the arrays. Has length equal to the number of arrays. Must be <code>NULL</code> if <code>ndups&gt;2</code>.</p></dd>


<dt>correlation</dt>
<dd><p>the inter-duplicate or inter-technical replicate correlation</p></dd>


<dt>contrasts</dt>
<dd><p>numeric matrix with rows corresponding to coefficients in <code>fit</code> and columns containing contrasts. May be a vector if there is only one contrast. <code>NA</code>s are not allowed.</p></dd>


<dt>proportion</dt>
<dd><p>numeric value between 0 and 1, assumed proportion of genes which are differentially expressed</p></dd>


<dt>stdev.coef.lim</dt>
<dd><p>numeric vector of length 2, assumed lower and upper limits for the standard deviation of log2-fold-changes for differentially expressed genes</p></dd>


<dt>robust</dt>
<dd><p>logical, should the estimation of <code>df.prior</code> and <code>var.prior</code> be robustified against outlier sample variances?</p></dd>


<dt>winsor.tail.p</dt>
<dd><p>numeric vector of length 1 or 2, giving left and right tail proportions of <code>x</code> to Winsorize. Used only when <code>robust=TRUE</code>.</p></dd>


<dt>coef</dt>
<dd><p>column number or column name specifying which coefficient or contrast of the linear model is of interest. For <code>topTable</code>, can also be a vector of column subscripts, in which case the gene ranking is by F-statistic for that set of contrasts.</p></dd>


<dt>adjust.method</dt>
<dd><p>method used to adjust the p-values for multiple testing.  Options, in increasing conservatism, include <code>"none"</code>, <code>"BH"</code>, <code>"BY"</code> and <code>"holm"</code>.
  See <code><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust</a></code> for the complete list of options. A <code>NULL</code> value will result in the default adjustment method, which is <code>"BH"</code>.</p></dd>


<dt>logger</dt>
<dd><p>a <code><a href="Logger.html">Logger</a></code> object.</p></dd>


<dt>...</dt>
<dd><p>further arguments to <code><a href="https://rdrr.io/pkg/limma/man/lmFit.html" class="external-link">lmFit</a></code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A list containing two elements:</p>


<dl><dt>statistic</dt>
<dd><p>A numeric vector, the values of the test statistic</p></dd>

<dt>significance</dt>
<dd><p>A numeric vector, the p-values of the selected test</p></dd>


</dl></div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Alessandro Barberis</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">#Seed</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1010</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#Define row/col size</span></span></span>
<span class="r-in"><span><span class="va">nr</span> <span class="op">=</span> <span class="fl">20</span></span></span>
<span class="r-in"><span><span class="va">nc</span> <span class="op">=</span> <span class="fl">20</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#Data</span></span></span>
<span class="r-in"><span><span class="va">x</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span></span></span>
<span class="r-in"><span> data <span class="op">=</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span>n <span class="op">=</span> <span class="va">nr</span><span class="op">*</span><span class="va">nc</span><span class="op">)</span>,</span></span>
<span class="r-in"><span> nrow <span class="op">=</span> <span class="va">nr</span>,</span></span>
<span class="r-in"><span> ncol <span class="op">=</span> <span class="va">nc</span>,</span></span>
<span class="r-in"><span> dimnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>   <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"g"</span>,<span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="va">nr</span><span class="op">)</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>   <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"S"</span>,<span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="va">nc</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span> <span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#Categorical output vector (binomial)</span></span></span>
<span class="r-in"><span><span class="va">y</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"wt"</span>,<span class="va">nc</span><span class="op">/</span><span class="fl">2</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"mut"</span>,<span class="va">nc</span><span class="op">/</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"S"</span>,<span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="va">nc</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="rowEBayesStatistics.html">rowEBayesStatistics</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x</span>,y<span class="op">=</span><span class="va">y</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $statistic</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1] 1.48314728 0.26127447 0.17149946 2.23891417 0.82910469 0.24320947</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [7] 2.81785282 1.14802510 0.44767056 0.04352112 0.02997531 0.92414587</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [13] 1.60206921 1.45753831 0.21663719 0.55468754 0.92130594 0.62506081</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [19] 0.02202450 1.13335883</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $significance</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1] 0.22692238 0.77006953 0.84240072 0.10657416 0.43643986 0.78410725</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [7] 0.05973406 0.31726271 0.63911520 0.95741234 0.97046949 0.39687025</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [13] 0.20147918 0.23280867 0.80522206 0.57425166 0.39799894 0.53522888</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [19] 0.97821627 0.32195006</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Alessandro Barberis.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

