<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="featscreen">
<title>Supervised screening methods • featscreen</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Supervised screening methods">
<meta property="og:description" content="featscreen">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">featscreen</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/featscreen.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Screening</h6>
    <a class="dropdown-item" href="../articles/supervised.html">Supervised screening methods</a>
    <a class="dropdown-item" href="../articles/unsupervised.html">Unupervised screening methods</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/alebarberis/featscreen/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Supervised screening methods</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/alebarberis/featscreen/blob/HEAD/vignettes/supervised.Rmd" class="external-link"><code>vignettes/supervised.Rmd</code></a></small>
      <div class="d-none name"><code>supervised.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>In the realm of data analysis, supervised feature screening emerges
as a pivotal process, guiding the selection of relevant features to
enhance predictive modeling and classification accuracy.</p>
<p><strong>featscreen</strong> empowers analysts and data scientists by
providing a comprehensive suite of ready-to-use functions tailored for
supervised feature screening.</p>
<p>From traditional correlation-based methods to sophisticated
statistical approaches such as Cox Proportional-Hazard models for
survival data, <strong>featscreen</strong> aims to bridge the gap
between complex methodologies and user-friendly implementation.</p>
<p>In this article we explore the different supervised feature screening
methods supported by <strong>featscreen</strong>.</p>
</div>
<div class="section level2">
<h2 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h2>
<p>Firstly, we load <code>featscreen</code> and other needed
packages:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/alebarberis/featscreen" class="external-link">featscreen</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'featscreen'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     mad, sd</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:graphics':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     screen</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="seed">Seed<a class="anchor" aria-label="anchor" href="#seed"></a>
</h2>
<p>Now we want to set a seed for the random number generation (RNG). In
fact, different R sessions have different seeds created from current
time and process ID by default, and consequently different simulation
results. By fixing a seed we ensure we will be able to reproduce the
results of this vignette. We can specify a seed by calling
<code><a href="https://rdrr.io/r/base/Random.html" class="external-link">?set.seed</a></code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Set a seed for RNG</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span></span>
<span>  <span class="co">#A seed</span></span>
<span>  seed <span class="op">=</span> <span class="fl">5381L</span>,                   <span class="co">#a randomly chosen integer value</span></span>
<span>  <span class="co">#The kind of RNG to use</span></span>
<span>  kind <span class="op">=</span> <span class="st">"Mersenne-Twister"</span>,      <span class="co">#we make explicit the current R default value</span></span>
<span>  <span class="co">#The kind of Normal generation</span></span>
<span>  normal.kind <span class="op">=</span> <span class="st">"Inversion"</span>       <span class="co">#we make explicit the current R default value</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="screening-methods">Screening Methods<a class="anchor" aria-label="anchor" href="#screening-methods"></a>
</h2>
<p>To view the list of currently supported supervised screening methods,
use the <code><a href="../reference/listAvailableScreeningMethods.html">?listAvailableScreeningMethods</a></code> function with the
<code>x</code> parameter set to <code>supervised</code>:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#list screening methods</span></span>
<span><span class="va">screening.methods</span> <span class="op">=</span> <span class="fu"><a href="../reference/listAvailableScreeningMethods.html">listAvailableScreeningMethods</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">'supervised'</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#print in table</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">screening.methods</span>, align <span class="op">=</span> <span class="st">'rc'</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="17%">
<col width="82%">
</colgroup>
<thead><tr class="header">
<th align="right">id</th>
<th align="center">name</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="right">pearson</td>
<td align="center">Pearson’s product moment correlation coefficient
t-test</td>
</tr>
<tr class="even">
<td align="right">spearman</td>
<td align="center">Spearman’s rank correlation coefficient t-test</td>
</tr>
<tr class="odd">
<td align="right">kendall</td>
<td align="center">Kendall’s rank correlation coefficient t-test</td>
</tr>
<tr class="even">
<td align="right">t.test.equal</td>
<td align="center">two-sample Student’s pooled t-test</td>
</tr>
<tr class="odd">
<td align="right">t.test.unequal</td>
<td align="center">two-sample t-test with the Welch modification to the
degrees of freedom</td>
</tr>
<tr class="even">
<td align="right">t.test.paired</td>
<td align="center">paired two-sample Student’s t-test</td>
</tr>
<tr class="odd">
<td align="right">w.test.ranksum</td>
<td align="center">two-sample Mann-Whitney U-test</td>
</tr>
<tr class="even">
<td align="right">w.test.paired</td>
<td align="center">paired two-sample Wilcoxon signed-rank test</td>
</tr>
<tr class="odd">
<td align="right">anova.equal</td>
<td align="center">one-way analysis of variance F-test</td>
</tr>
<tr class="even">
<td align="right">anova.unequal</td>
<td align="center">one-way analysis of variance F-test with Welch
correction</td>
</tr>
<tr class="odd">
<td align="right">kruskal.wallis</td>
<td align="center">Kruskal-Wallis H-test</td>
</tr>
<tr class="even">
<td align="right">chisq.test</td>
<td align="center">Pearson’s χ²-test</td>
</tr>
<tr class="odd">
<td align="right">coxph</td>
<td align="center">Cox PH regression coefficient z-test</td>
</tr>
<tr class="even">
<td align="right">moderated.t</td>
<td align="center">empirical Bayes moderated t-test</td>
</tr>
<tr class="odd">
<td align="right">moderated.F</td>
<td align="center">empirical Bayes moderated F-test</td>
</tr>
<tr class="even">
<td align="right">sam.test</td>
<td align="center">significant analysis of microarrays permutation
test</td>
</tr>
</tbody>
</table>
<div class="section level3">
<h3 id="correlation">Correlation<a class="anchor" aria-label="anchor" href="#correlation"></a>
</h3>
<p>Correlation analysis plays a fundamental role in understanding
relationships between variables in the context of supervised feature
screening. It provides a quantitative measure of the strength and
direction of associations, helping identify features that may contribute
significantly to predictive models.</p>
<p><strong>featscreen</strong> supports various correlation methods,
each catering to different data characteristics and analytical
needs.</p>
<div class="section level4">
<h4 id="pearsons-correlation">Pearson’s Correlation<a class="anchor" aria-label="anchor" href="#pearsons-correlation"></a>
</h4>
<p><strong>Pearson’s correlation</strong> is a widely-used measure that
assesses linear relationships between features and the target variable.
Represented by the correlation coefficient <em>r</em>, this method
quantifies the extent to which two variables change together. The
coefficient ranges from -1 to 1, where -1 indicates a perfect negative
linear relationship, 1 denotes a perfect positive linear relationship,
and 0 signifies no linear correlation. Pearson’s correlation is
sensitive to outliers and assumes a linear relationship between
variables.</p>
<p>Pearson’s correlation is computed in <strong>featscreen</strong> by
calling the function <code><a href="../reference/rowPearsonCor.html">?rowPearsonCor</a></code>. From the
documentation, we can see that this is a wrapper function that
internally calls <code>?row_cor_pearson</code> from the
<code>matrixTests</code> package.</p>
<p>Let’s explore a practical example to illustrate the application of
Pearson’s correlation.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Data</span></span>
<span><span class="va">x</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">10</span> <span class="op">*</span> <span class="fl">20</span><span class="op">)</span>, <span class="fl">10</span>, <span class="fl">20</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">20</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Compute</span></span>
<span><span class="fu"><a href="../reference/rowPearsonCor.html">rowPearsonCor</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span><span class="op">)</span></span>
<span><span class="co">#&gt; $statistic</span></span>
<span><span class="co">#&gt;  [1]  1.55424671 -3.21135119 -0.48824958 -0.62940963  0.01709722 -1.30914381</span></span>
<span><span class="co">#&gt;  [7] -1.07986577  1.11138904 -0.42259796  0.39229005</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $significance</span></span>
<span><span class="co">#&gt;  [1] 0.137531731 0.004840161 0.631267786 0.536988472 0.986547148 0.206952421</span></span>
<span><span class="co">#&gt;  [7] 0.294459519 0.281026013 0.677593887 0.699450166</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="spearmans-correlation">Spearman’s Correlation<a class="anchor" aria-label="anchor" href="#spearmans-correlation"></a>
</h4>
<p><strong>Spearman’s correlation</strong> is a non-parametric measure
that evaluates monotonic relationships between variables. By ranking the
data values and assessing the correlation of the ranks, Spearman’s
correlation provides a robust alternative to Pearson’s correlation,
especially in the presence of outliers or non-linear patterns. The
resulting coefficient, denoted by <em>rho</em>, ranges between -1 and 1,
where extreme values indicate strong monotonic relationships.</p>
<p>In <strong>featscreen</strong>, Spearman’s correlation is computed by
calling the function <code><a href="../reference/rowSpearmanCor.html">?rowSpearmanCor</a></code>. From the
documentation, we can see that this is a wrapper function that
internally calls <code><a href="https://rdrr.io/r/stats/cor.test.html" class="external-link">?cor.test</a></code> from the <code>stats</code>
package.</p>
<p>Let see an example.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Data</span></span>
<span><span class="va">x</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">10</span> <span class="op">*</span> <span class="fl">20</span><span class="op">)</span>, <span class="fl">10</span>, <span class="fl">20</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">20</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Compute</span></span>
<span><span class="fu"><a href="../reference/rowSpearmanCor.html">rowSpearmanCor</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span><span class="op">)</span></span>
<span><span class="co">#&gt; $statistic</span></span>
<span><span class="co">#&gt;  [1] 2004 1418 1582 1022 1114 1254 1376  770 1502 1542</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $significance</span></span>
<span><span class="co">#&gt;  [1] 0.02410399 0.78197196 0.42194872 0.32438799 0.49229843 0.81146975</span></span>
<span><span class="co">#&gt;  [7] 0.88623825 0.06578794 0.58563682 0.50046206</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="kendalls-correlation">Kendall’s Correlation<a class="anchor" aria-label="anchor" href="#kendalls-correlation"></a>
</h4>
<p><strong>Kendall’s correlation</strong> is another non-parametric
method for assessing correlation. Similar to Spearman’s correlation, it
is based on the ranking of data. It measures the similarity in the
ordering of data pairs between two variables, making it suitable for
ordinal data and less sensitive to outliers. Kendall’s coefficient,
denoted by <em>tau</em>, ranges from -1 to 1, with negative values
indicating a negative association, positive values signifying a positive
association, and 0 representing no association.</p>
<p>In <strong>featscreen</strong>, Kendall’s correlation is computed by
calling the function <code><a href="../reference/rowKendallCor.html">?rowKendallCor</a></code>. From the
documentation, we can see that this is a wrapper function that
internally calls <code><a href="https://rdrr.io/r/stats/cor.test.html" class="external-link">?cor.test</a></code> from the <code>stats</code>
package.</p>
<p>Let see an example.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Data</span></span>
<span><span class="va">x</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">10</span> <span class="op">*</span> <span class="fl">20</span><span class="op">)</span>, <span class="fl">10</span>, <span class="fl">20</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">20</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Compute</span></span>
<span><span class="fu"><a href="../reference/rowKendallCor.html">rowKendallCor</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span><span class="op">)</span></span>
<span><span class="co">#&gt; $statistic</span></span>
<span><span class="co">#&gt;  [1]  88  81  73 121  77 112  74 117  63  96</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $significance</span></span>
<span><span class="co">#&gt;  [1] 0.67710824 0.38585715 0.16497641 0.09833022 0.25983999 0.28837820</span></span>
<span><span class="co">#&gt;  [7] 0.18588304 0.16497641 0.03976214 0.97446690</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="two-sample-tests">Two-Sample Tests<a class="anchor" aria-label="anchor" href="#two-sample-tests"></a>
</h3>
<p>Two-sample tests are indispensable tools in supervised feature
screening, offering insights into the differences between two groups of
data.</p>
<p><strong>featscreen</strong> incorporates a range of two-sample tests,
each catering to diverse scenarios and assumptions. Whether examining
mean differences with parametric tests or assessing variations in
medians with non-parametric tests, these methods play a pivotal role in
uncovering significant distinctions between groups, contributing to the
identification of influential features within the feature space.</p>
<div class="section level4">
<h4 id="two-sample-students-t-test">Two-Sample Student’s t-Test<a class="anchor" aria-label="anchor" href="#two-sample-students-t-test"></a>
</h4>
<p>The <strong>Two-Sample Student’s t-Test</strong> is a parametric
method used to assess whether the means of two groups are significantly
different. It is applicable when the data follow a normal distribution.
The t-test results in a t-statistic and a p-value, helping analysts
determine the statistical significance of observed differences.</p>
<p>Two-sample t-tests can be further divided into:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Unpaired</strong> two-samples t-tests</li>
<li>
<strong>Paired</strong> two-samples t-tests</li>
</ol>
<div class="section level5">
<h5 id="unpaired-samples">Unpaired Samples<a class="anchor" aria-label="anchor" href="#unpaired-samples"></a>
</h5>
<p>T-tests provide invaluable insights into differences between two
independent groups.</p>
<p><strong>featscreen</strong> supports unpaired two-sample t-tests,
offering flexibility to handle scenarios with both equal and unequal
variances. This class of t-tests is particularly applicable when
comparing means between two groups where each observation in one group
is independent of the other.</p>
<div class="section level6">
<h6 id="equal-variance-t-test">Equal Variance T-Test<a class="anchor" aria-label="anchor" href="#equal-variance-t-test"></a>
</h6>
<p>The <strong>Equal Variance T-Test</strong>, often referred to as the
traditional or pooled t-test, assumes that the variances of the two
groups being compared are equal. This test is suitable when the data in
both groups follow a normal distribution and exhibit homogeneity of
variances.</p>
<p><strong>featscreen</strong> supports the Equal Variance T-Test
through the <code><a href="../reference/rowEqualVarT.html">?rowEqualVarT</a></code> function.</p>
<p>From the documentation, we can see that this is a wrapper function
that internally calls <code>?row_t_equalvar</code> from the
<code>matrixTests</code> package.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Data</span></span>
<span><span class="va">x</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">10</span> <span class="op">*</span> <span class="fl">20</span><span class="op">)</span>, <span class="fl">10</span>, <span class="fl">20</span><span class="op">)</span></span>
<span><span class="co"># Assuming 'group1' and 'group2' are indicating independent measurements</span></span>
<span><span class="va">g</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">'group1'</span>,<span class="fl">10</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">'group2'</span>,<span class="fl">10</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Compute</span></span>
<span><span class="fu"><a href="../reference/rowEqualVarT.html">rowEqualVarT</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, g <span class="op">=</span> <span class="va">g</span><span class="op">)</span></span>
<span><span class="co">#&gt; $statistic</span></span>
<span><span class="co">#&gt;  [1]  0.004667386 -0.600050691 -0.524350283 -0.158893662 -1.482498596</span></span>
<span><span class="co">#&gt;  [6] -0.461348953 -0.034249081 -0.471759890  0.923177678 -1.557724376</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $significance</span></span>
<span><span class="co">#&gt;  [1] 0.9963273 0.5559521 0.6064326 0.8755219 0.1555043 0.6500745 0.9730555</span></span>
<span><span class="co">#&gt;  [8] 0.6427664 0.3681249 0.1367066</span></span></code></pre></div>
</div>
<div class="section level6">
<h6 id="unequal-variance-t-test">Unequal Variance T-Test<a class="anchor" aria-label="anchor" href="#unequal-variance-t-test"></a>
</h6>
<p>The <strong>Unequal Variance T-Test</strong>, also known as
<em>Welch’s T-Test</em>, is employed when the assumption of equal
variances is not met. This test is robust in scenarios where variances
in the two groups differ.</p>
<p>The Unequal Variance T-Test is available in
<strong>featscreen</strong> through the <code><a href="../reference/rowUnequalVarT.html">?rowUnequalVarT</a></code>
function.</p>
<p>From the documentation, we can see that this is a wrapper function
that internally calls <code>?row_t_welch</code> from the
<code>matrixTests</code> package.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Data</span></span>
<span><span class="va">x</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">10</span> <span class="op">*</span> <span class="fl">10</span>, mean <span class="op">=</span> <span class="fl">1</span>, sd <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>, <span class="fl">10</span>, <span class="fl">10</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">10</span> <span class="op">*</span> <span class="fl">10</span>, mean <span class="op">=</span> <span class="fl">3</span>, sd <span class="op">=</span> <span class="fl">4</span><span class="op">)</span>, <span class="fl">10</span>, <span class="fl">10</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Assuming 'group1' and 'group2' are indicating independent measurements</span></span>
<span><span class="va">g</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">'group1'</span>,<span class="fl">10</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">'group2'</span>,<span class="fl">10</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Compute</span></span>
<span><span class="fu"><a href="../reference/rowUnequalVarT.html">rowUnequalVarT</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, g <span class="op">=</span> <span class="va">g</span><span class="op">)</span></span>
<span><span class="co">#&gt; $statistic</span></span>
<span><span class="co">#&gt;  [1] -2.2405783 -2.2538213  0.1452079  0.9806682 -2.9989363 -0.8862154</span></span>
<span><span class="co">#&gt;  [7] -3.6757116 -1.3743303 -2.4372740 -6.4725307</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $significance</span></span>
<span><span class="co">#&gt;  [1] 4.587359e-02 4.293420e-02 8.870308e-01 3.473898e-01 1.174654e-02</span></span>
<span><span class="co">#&gt;  [6] 3.894469e-01 3.692485e-03 1.946094e-01 2.910688e-02 4.101379e-05</span></span></code></pre></div>
</div>
</div>
<div class="section level5">
<h5 id="paired-samples">Paired Samples<a class="anchor" aria-label="anchor" href="#paired-samples"></a>
</h5>
<p>The <strong>paired t-test</strong> variant is available for
situations where measurements are paired, such as repeated measurements
on the same subjects or matched pairs of subjects. This variant of the
t-test assesses whether the mean difference between paired observations
is statistically different from zero.</p>
<p>The paired t-test is particularly useful in experimental designs
where each subject is subjected to two different conditions, treatments,
or time points, and the goal is to determine whether there is a
significant change within each pair. For instance, in clinical trials,
the paired t-test may be applied to assess the efficacy of a treatment
by comparing measurements taken before and after the treatment for the
same group of individuals.</p>
<p>In <strong>featscreen</strong>, a paired t-test is computed by
calling the function <code><a href="../reference/rowPairedT.html">?rowPairedT</a></code>. From the documentation,
we can see that this is a wrapper function that internally calls
<code>?row_t_paired</code> from the <code>matrixTests</code>
package.</p>
<p>Let see an example.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Data</span></span>
<span><span class="va">x</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">10</span> <span class="op">*</span> <span class="fl">20</span><span class="op">)</span>, <span class="fl">10</span>, <span class="fl">20</span><span class="op">)</span></span>
<span><span class="co"># Assuming 'before' and 'after' are indicating paired measurements</span></span>
<span><span class="va">g</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">'before'</span>,<span class="fl">10</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">'after'</span>,<span class="fl">10</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Compute</span></span>
<span><span class="fu"><a href="../reference/rowPairedT.html">rowPairedT</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, g <span class="op">=</span> <span class="va">g</span><span class="op">)</span></span>
<span><span class="co">#&gt; $statistic</span></span>
<span><span class="co">#&gt;  [1] -0.70249807  0.24745281 -0.37887763 -0.26889905 -2.09181818  0.61294897</span></span>
<span><span class="co">#&gt;  [7]  0.04841629 -0.04692657 -0.34187150 -0.31344754</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $significance</span></span>
<span><span class="co">#&gt;  [1] 0.50013315 0.81011060 0.71356227 0.79407000 0.06598845 0.55508069</span></span>
<span><span class="co">#&gt;  [7] 0.96244188 0.96359655 0.74029490 0.76108331</span></span></code></pre></div>
<p>In this example, the paired t-test is applied to assess whether there
is a significant difference in measurements taken before and after a
treatment or intervention. Users can adapt this approach to various
domains, providing a powerful means of exploring paired data within the
supervised feature screening process.</p>
</div>
</div>
<div class="section level4">
<h4 id="two-sample-wilcoxons-signed-rank-test">Two-Sample Wilcoxon’s Signed-Rank Test<a class="anchor" aria-label="anchor" href="#two-sample-wilcoxons-signed-rank-test"></a>
</h4>
<p>The <strong>Two-Sample Wilcoxon’s Signed-Rank Test</strong> is a
non-parametric alternative to the t-test, suitable for situations where
the assumption of normality is violated. This test assesses whether the
distribution of differences between paired observations differs
significantly from zero. It is particularly robust against outliers and
variations in the shape of the distribution, making it a valuable option
for scenarios with non-normally distributed data.</p>
<p>In <strong>featscreen</strong>, the wilcoxon signed-rank test is
computed by calling the function <code><a href="../reference/rowPairedWilcoxonT.html">?rowPairedWilcoxonT</a></code>.</p>
<p>From the documentation, we can see that this is a wrapper function
that internally calls <code>?row_wilcoxon_paired</code> from the
<code>matrixTests</code> package.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Data</span></span>
<span><span class="va">x</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">10</span> <span class="op">*</span> <span class="fl">20</span><span class="op">)</span>, <span class="fl">10</span>, <span class="fl">20</span><span class="op">)</span></span>
<span><span class="co"># Assuming 'before' and 'after' are indicating paired measurements</span></span>
<span><span class="va">g</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">'before'</span>,<span class="fl">10</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">'after'</span>,<span class="fl">10</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Compute</span></span>
<span><span class="fu"><a href="../reference/rowWilcoxonT.html">rowWilcoxonT</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, g <span class="op">=</span> <span class="va">g</span><span class="op">)</span></span>
<span><span class="co">#&gt; $statistic</span></span>
<span><span class="co">#&gt;  [1] 82 54 74 65 45 45 50 30 51 61</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $significance</span></span>
<span><span class="co">#&gt;  [1] 0.01468964 0.79593626 0.07525601 0.27986101 0.73936435 0.73936435</span></span>
<span><span class="co">#&gt;  [7] 1.00000000 0.14314014 0.97051246 0.43587218</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="two-sample-mann-whitney-u-test">Two-Sample Mann-Whitney U-Test<a class="anchor" aria-label="anchor" href="#two-sample-mann-whitney-u-test"></a>
</h4>
<p>The <strong>Two-Sample Mann-Whitney U-Test</strong> (also known as
<em>Wilcoxon Rank-Sum Test</em>) extends the Wilcoxon test to
independent samples, allowing analysts to evaluate whether the
distributions of two groups differ significantly. This non-parametric
test is suitable for situations where assumptions of normality are not
met, providing an effective tool for feature screening in diverse
datasets. The Mann-Whitney U-Test results in a U-statistic and a
p-value, aiding in the determination of statistical significance between
groups.</p>
<p>In <strong>featscreen</strong>, the Wilcoxon rank-sum test is
computed by calling the function <code><a href="../reference/rowWilcoxonT.html">?rowWilcoxonT</a></code>.</p>
<p>From the documentation, we can see that this is a wrapper function
that internally calls <code>?row_wilcoxon_twosample</code> from the
<code>matrixTests</code> package.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Data</span></span>
<span><span class="va">x</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">10</span> <span class="op">*</span> <span class="fl">20</span><span class="op">)</span>, <span class="fl">10</span>, <span class="fl">20</span><span class="op">)</span></span>
<span><span class="co"># Assuming 'before' and 'after' are indicating independent measurements</span></span>
<span><span class="va">g</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">'before'</span>,<span class="fl">10</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">'after'</span>,<span class="fl">10</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Compute</span></span>
<span><span class="fu"><a href="../reference/rowWilcoxonT.html">rowWilcoxonT</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, g <span class="op">=</span> <span class="va">g</span><span class="op">)</span></span>
<span><span class="co">#&gt; $statistic</span></span>
<span><span class="co">#&gt;  [1] 43 60 56 55 15 54 59 67 51 47</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $significance</span></span>
<span><span class="co">#&gt;  [1] 0.630528914 0.481250947 0.684210526 0.739364351 0.006841456 0.795936262</span></span>
<span><span class="co">#&gt;  [7] 0.528848860 0.217562623 0.970512460 0.853428305</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="comparing-multiple-groups">Comparing Multiple Groups<a class="anchor" aria-label="anchor" href="#comparing-multiple-groups"></a>
</h3>
<p>Comparing multiple groups is a crucial aspect of supervised feature
screening, allowing analysts to discern variations in feature behavior
across different categorical or experimental conditions.</p>
<p><strong>featscreen</strong> integrates diverse statistical methods
tailored for such comparisons, accommodating scenarios with varying
assumptions and data characteristics. Whether exploring mean differences
with ANOVA, assessing median disparities with non-parametric tests, or
examining categorical associations with the Chi-Square Test,
<strong>featscreen</strong> provides a comprehensive suite of tools to
unravel nuanced patterns within the feature space.</p>
<div class="section level4">
<h4 id="one-way-analysis-of-variance-f-test">One-way Analysis of Variance F-Test<a class="anchor" aria-label="anchor" href="#one-way-analysis-of-variance-f-test"></a>
</h4>
<p>The One-way Analysis of Variance (ANOVA) F-Test stands as a
fundamental parametric method designed for comparing means across two or
more independent groups. This statistical test addresses the question of
whether there are significant differences in the means of distinct
groups, making it a versatile tool for analyzing experimental data with
multiple categorical conditions or levels. ANOVA relies on the
assumption of normality and homogeneity of variances among the
groups.</p>
<p>The F-Test results in an F-statistic and a p-value, enabling analysts
to make informed decisions about the statistical significance of
observed differences in means.</p>
<p><strong>featscreen</strong> supports ANOVA with equal or unequal
variance.</p>
<div class="section level5">
<h5 id="equal-variance-anova">Equal Variance ANOVA<a class="anchor" aria-label="anchor" href="#equal-variance-anova"></a>
</h5>
<p>The Equal Variance ANOVA is used to assess mean differences among
multiple independent groups. This variant of ANOVA, also known as the
traditional or pooled ANOVA, assumes homogeneity of variances and
normality within each group.</p>
<p>In <strong>featscreen</strong>, the pooled one-way analysis of
variance test is computed by calling the function
<code><a href="../reference/rowEqualVarOneWayAnova.html">?rowEqualVarOneWayAnova</a></code>.</p>
<p>From the documentation, we can see that this is a wrapper function
that internally calls <code>?row_oneway_equalvar</code> from the
<code>matrixTests</code> package.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Data</span></span>
<span><span class="va">x</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">10</span> <span class="op">*</span> <span class="fl">20</span><span class="op">)</span>, <span class="fl">10</span>, <span class="fl">20</span><span class="op">)</span></span>
<span><span class="co"># Assuming 'before' and 'after' are indicating independent measurements</span></span>
<span><span class="va">g</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">'before'</span>,<span class="fl">10</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">'after'</span>,<span class="fl">10</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Compute</span></span>
<span><span class="fu"><a href="../reference/rowEqualVarOneWayAnova.html">rowEqualVarOneWayAnova</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, g <span class="op">=</span> <span class="va">g</span><span class="op">)</span></span>
<span><span class="co">#&gt; $statistic</span></span>
<span><span class="co">#&gt;  [1] 10.19316848  0.03225024  0.04230048  0.06552166  2.00635238  0.59021632</span></span>
<span><span class="co">#&gt;  [7]  0.01912477  0.42145581  0.14022777  0.11147536</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $significance</span></span>
<span><span class="co">#&gt;  [1] 0.005043041 0.859485661 0.839356234 0.800874877 0.173715440 0.452297572</span></span>
<span><span class="co">#&gt;  [7] 0.891544743 0.524409445 0.712429816 0.742328966</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="unequal-variance-anova">Unequal Variance ANOVA<a class="anchor" aria-label="anchor" href="#unequal-variance-anova"></a>
</h5>
<p>The Unequal Variance ANOVA, also known as Welch’s ANOVA, accommodates
scenarios where the variances among groups differ, offering a robust
approach when homogeneity of variances cannot be assumed.</p>
<p>In <strong>featscreen</strong>, the one-way analysis of variance test
with Welch correction is computed by calling the function
<code><a href="../reference/rowUnequalVarOneWayAnova.html">?rowUnequalVarOneWayAnova</a></code>.</p>
<p>From the documentation, we can see that this is a wrapper function
that internally calls <code>?row_oneway_welch</code> from the
<code>matrixTests</code> package.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Data</span></span>
<span><span class="va">x</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">10</span> <span class="op">*</span> <span class="fl">10</span>, mean <span class="op">=</span> <span class="fl">1</span>, sd <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>, <span class="fl">10</span>, <span class="fl">10</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">10</span> <span class="op">*</span> <span class="fl">10</span>, mean <span class="op">=</span> <span class="fl">3</span>, sd <span class="op">=</span> <span class="fl">4</span><span class="op">)</span>, <span class="fl">10</span>, <span class="fl">10</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Assuming 'before' and 'after' are indicating independent measurements</span></span>
<span><span class="va">g</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">'before'</span>,<span class="fl">10</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">'after'</span>,<span class="fl">10</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Compute</span></span>
<span><span class="fu"><a href="../reference/rowUnequalVarOneWayAnova.html">rowUnequalVarOneWayAnova</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, g <span class="op">=</span> <span class="va">g</span><span class="op">)</span></span>
<span><span class="co">#&gt; $statistic</span></span>
<span><span class="co">#&gt;  [1] 0.006616130 4.551030128 3.854863607 0.708377597 1.118078432 0.224635260</span></span>
<span><span class="co">#&gt;  [7] 1.821935578 0.001472292 0.821963664 0.015927117</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $significance</span></span>
<span><span class="co">#&gt;  [1] 0.93674061 0.05482032 0.06530365 0.41674366 0.31000755 0.64494599</span></span>
<span><span class="co">#&gt;  [7] 0.19982402 0.96992567 0.38033507 0.90118277</span></span></code></pre></div>
</div>
</div>
<div class="section level4">
<h4 id="kruskal-wallis-test">Kruskal-Wallis Test<a class="anchor" aria-label="anchor" href="#kruskal-wallis-test"></a>
</h4>
<p>A non-parametric alternative to ANOVA, the Kruskal-Wallis test is
applied when the assumption of normality is not met or when analyzing
ordinal data.</p>
<p>This test evaluates whether the distributions of two or more
independent groups differ significantly. The Kruskal-Wallis Test results
in a chi-square statistic and a p-value, offering a robust approach for
scenarios where parametric assumptions may be violated.</p>
<p>In <strong>featscreen</strong>, the Kruskal-Wallis H-test is computed
by calling the function <code><a href="../reference/rowKruskalWallis.html">?rowKruskalWallis</a></code>.</p>
<p>From the documentation, we can see that this is a wrapper function
that internally calls <code>?row_kruskalwallis</code> from the
<code>matrixTests</code> package.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Data</span></span>
<span><span class="va">x</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">10</span> <span class="op">*</span> <span class="fl">10</span>, mean <span class="op">=</span> <span class="fl">1</span>, sd <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>, <span class="fl">10</span>, <span class="fl">10</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">10</span> <span class="op">*</span> <span class="fl">10</span>, mean <span class="op">=</span> <span class="fl">3</span>, sd <span class="op">=</span> <span class="fl">4</span><span class="op">)</span>, <span class="fl">10</span>, <span class="fl">10</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Assuming 'before' and 'after' are indicating independent measurements</span></span>
<span><span class="va">g</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">'before'</span>,<span class="fl">10</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">'after'</span>,<span class="fl">10</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Compute</span></span>
<span><span class="fu"><a href="../reference/rowUnequalVarOneWayAnova.html">rowUnequalVarOneWayAnova</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, g <span class="op">=</span> <span class="va">g</span><span class="op">)</span></span>
<span><span class="co">#&gt; $statistic</span></span>
<span><span class="co">#&gt;  [1]  1.071701  3.756477  1.446872  1.520712  0.582938  2.292538  1.062549</span></span>
<span><span class="co">#&gt;  [8]  5.109783  6.530679 16.376850</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $significance</span></span>
<span><span class="co">#&gt;  [1] 0.325304467 0.073453691 0.253354235 0.239221847 0.456458177 0.160090799</span></span>
<span><span class="co">#&gt;  [7] 0.320451489 0.041018406 0.027682348 0.001356772</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="chi-square-test">Chi-Square Test<a class="anchor" aria-label="anchor" href="#chi-square-test"></a>
</h4>
<p>Chi-Square Test investigates associations between categorical
variables, providing a statistical basis for feature selection in
contingency tables.</p>
<p>In <strong>featscreen</strong>, the Pearson’s Chi-squared test is
computed by calling the function <code><a href="../reference/rowPearsonChiSq.html">?rowPearsonChiSq</a></code>.</p>
<p>From the documentation, we can see that this is a wrapper function
that internally calls <code><a href="https://rdrr.io/r/stats/chisq.test.html" class="external-link">?chisq.test</a></code> from the
<code>stats</code> package.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Data</span></span>
<span><span class="co">## Matrix with 2 features: 'mutation status' and 'sex'</span></span>
<span><span class="va">x</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span></span>
<span>   <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"mut"</span>, <span class="st">"wt"</span><span class="op">)</span>,<span class="fl">30</span>,<span class="cn">TRUE</span><span class="op">)</span>, <span class="fl">1</span>, <span class="fl">30</span><span class="op">)</span>,</span>
<span>   <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"m"</span>, <span class="st">"f"</span><span class="op">)</span>   ,<span class="fl">30</span>,<span class="cn">TRUE</span><span class="op">)</span>, <span class="fl">1</span>, <span class="fl">30</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">g</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"a"</span>,<span class="st">"b"</span>,<span class="st">"c"</span><span class="op">)</span>, <span class="fl">30</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Compute</span></span>
<span><span class="fu"><a href="../reference/rowPearsonChiSq.html">rowPearsonChiSq</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, g <span class="op">=</span> <span class="va">g</span>, simulate.p.value <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; $statistic</span></span>
<span><span class="co">#&gt; [1] 1.574598 1.715343</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $significance</span></span>
<span><span class="co">#&gt; [1] 0.5192404 0.4582709</span></span></code></pre></div>
<p>The example showcases the application of the Chi-Square Test using a
contingency table with two categorical features: ‘mutation status’ and
‘sex’. The test results include a Chi-Square statistic and a p-value,
providing analysts with insights into the significance of associations
between these categorical variables across different groups.</p>
</div>
</div>
<div class="section level3">
<h3 id="survival-data">Survival Data<a class="anchor" aria-label="anchor" href="#survival-data"></a>
</h3>
<p>Survival data analysis occupies a distinctive realm within feature
screening, focusing on time-to-event outcomes. In many fields, including
medical research and epidemiology, understanding the duration until a
particular event occurs is crucial.</p>
<div class="section level4">
<h4 id="cox-ph-regression-coefficient-z-test">Cox PH Regression Coefficient z-test<a class="anchor" aria-label="anchor" href="#cox-ph-regression-coefficient-z-test"></a>
</h4>
<p>Survival analysis often employs Cox Proportional-Hazard (PH) models,
a powerful statistical approach for understanding the impact of
covariates on the time to an event.</p>
<p>In <strong>featscreen</strong>, the Cox PH Regression Coefficient
Z-Test is used for assessing the significance of regression coefficients
in the context of survival data.</p>
<p>Within <strong>featscreen</strong>, the Cox Proportional-Hazard
regression coefficient Z-Test is executed through the
<code><a href="../reference/rowCoxPH.html">?rowCoxPH</a></code> function. From the documentation, we can see that
this this function serves as a wrapper, internally calling
<code>?coxph</code> from the <code>survival</code> package.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Data</span></span>
<span><span class="va">x</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">10</span> <span class="op">*</span> <span class="fl">7</span><span class="op">)</span>, <span class="fl">10</span>, <span class="fl">7</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>   time <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>,<span class="fl">3</span>,<span class="fl">1</span>,<span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">2</span>,<span class="fl">3</span><span class="op">)</span>,</span>
<span>   status <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span>,<span class="fl">1</span>,<span class="fl">0</span>,<span class="fl">1</span>,<span class="fl">1</span>,<span class="fl">0</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Compute</span></span>
<span><span class="fu"><a href="../reference/rowCoxPH.html">rowCoxPH</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span><span class="op">)</span></span>
<span><span class="co">#&gt; $statistic</span></span>
<span><span class="co">#&gt;  [1] 4.6493152 0.8127091 1.5519850 2.2382765 0.6912241 0.4545558 2.7561108</span></span>
<span><span class="co">#&gt;  [8] 1.4729205 0.4579275 1.4901095</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $significance</span></span>
<span><span class="co">#&gt;  [1] 0.05461149 0.77957911 0.44449927 0.16312546 0.46587119 0.17767696</span></span>
<span><span class="co">#&gt;  [7] 0.11668917 0.53995856 0.29448808 0.47456875</span></span></code></pre></div>
<p>The example above showcases the application of the Cox PH Regression
Coefficient Z-Test in <strong>featscreen</strong>, using simulated
survival data. The results include z-statistics and p-values, aiding
analysts in discerning the significance of individual features in
influencing time-to-event outcomes.</p>
</div>
</div>
<div class="section level3">
<h3 id="bio-statistical-functions">Bio-statistical Functions<a class="anchor" aria-label="anchor" href="#bio-statistical-functions"></a>
</h3>
<p>In the pursuit of understanding complex biological data, there have
been continuous efforts to refine statistical techniques, aiming to
better model the intricacies inherent in bioinformatic datasets.</p>
<p>These bio-statistical functions can play a crucial role in feature
screening.</p>
<div class="section level4">
<h4 id="empirical-bayes-moderated-tests">Empirical Bayes Moderated Tests<a class="anchor" aria-label="anchor" href="#empirical-bayes-moderated-tests"></a>
</h4>
<p>These tests offer a robust approach to detecting differentially
expressed features in gene expression data arising from microarray or
RNA-seq technologies.</p>
<p>Particularly, here we take advantage of the <code>limma</code> R
package. Renowned for its effectiveness in gene expression analysis,
<code>limma</code> employs linear modeling to fit a model to the
systematic part of the data, paving the way for robust statistical
inference.</p>
<p>The basic statistic used for significance analysis is the moderated
t-statistic, which has the same interpretation as an ordinary
t-statistic except that the standard errors have been moderated across
genes, i.e., squeezed towards a common value, using a simple Bayesian
model. As reported in the <a href="https://bioconductor.org/packages/devel/bioc/vignettes/limma/inst/doc/usersguide.pdf" class="external-link">limma
user guide</a>:</p>
<blockquote>
<p>Moderated t-statistics lead to p-values in the same way that ordinary
t-statistics do except that the degrees of freedom are increased,
reflecting the greater reliability associated with the smoothed standard
errors. The effectiveness of the moderated t approach has been
demonstrated on test data sets for which the differential expression
status of each probe is known.</p>
</blockquote>
<p>In <strong>featscreen</strong>, the empirical Bayes moderated
t-statistics are computed by calling the function
<code><a href="../reference/rowEBayesStatistics.html">?rowEBayesStatistics</a></code>.</p>
<p>From the documentation, we can see that this is a wrapper function
that internally calls different functions from the <code>limma</code>
package:</p>
<ol style="list-style-type: decimal">
<li>
<code>voom</code>: (optional) transform count data for linear
modelling.</li>
<li>
<code>lmFit</code>: linear model fitting for each feature.</li>
<li>
<code>contrasts.fit</code>: compute coefficients and standard errors
for a given set of contrasts.</li>
<li>
<code>eBayes</code>: compute moderated moderated
t-/F-statistic.</li>
<li>
<code>topTable</code>: summarise the linear model fit.</li>
</ol>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Define row/col size</span></span>
<span><span class="va">nr</span> <span class="op">=</span> <span class="fl">20</span></span>
<span><span class="va">nc</span> <span class="op">=</span> <span class="fl">20</span></span>
<span></span>
<span><span class="co">#Data</span></span>
<span><span class="va">x</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span>n <span class="op">=</span> <span class="va">nr</span><span class="op">*</span><span class="va">nc</span><span class="op">)</span>,</span>
<span>  nrow <span class="op">=</span> <span class="va">nr</span>,</span>
<span>  ncol <span class="op">=</span> <span class="va">nc</span>,</span>
<span>  dimnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"g"</span>,<span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="va">nr</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"S"</span>,<span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="va">nc</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#Categorical output vector (multinomial)</span></span>
<span><span class="va">y</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"I"</span>,<span class="st">"II"</span>,<span class="st">"III"</span><span class="op">)</span>,size<span class="op">=</span><span class="va">nc</span>,replace<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"S"</span>,<span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="va">nc</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Compute</span></span>
<span><span class="fu"><a href="../reference/rowEBayesStatistics.html">rowEBayesStatistics</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x</span>,y<span class="op">=</span><span class="va">y</span><span class="op">)</span></span>
<span><span class="co">#&gt; $statistic</span></span>
<span><span class="co">#&gt;  [1] 1.0137223 0.4925825 1.9615795 0.3904325 0.5573581 3.5581814 1.5763334</span></span>
<span><span class="co">#&gt;  [8] 0.4406250 0.2684814 1.2074018 0.4657751 0.7166232 0.2134995 0.3921002</span></span>
<span><span class="co">#&gt; [15] 1.4700849 0.8278437 0.5124011 1.1684151 0.6404147 0.5607559</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $significance</span></span>
<span><span class="co">#&gt;  [1] 0.38532148 0.68741566 0.11735461 0.75989623 0.64316005 0.01362243</span></span>
<span><span class="co">#&gt;  [7] 0.19275114 0.72394529 0.84816445 0.30525490 0.70616183 0.54188990</span></span>
<span><span class="co">#&gt; [13] 0.88710173 0.75869384 0.22043643 0.47827473 0.67371081 0.32008245</span></span>
<span><span class="co">#&gt; [19] 0.58891259 0.64088326</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="sam-permutation-test">SAM Permutation Test<a class="anchor" aria-label="anchor" href="#sam-permutation-test"></a>
</h4>
<p>Significance Analysis of Microarrays (SAM) permutation test is a
powerful tool for identifying features with significant expression
changes in microarray data. It was proposed by Tusher, Tibshirani and
Chu <span class="citation">(Tusher, Tibshirani, and Chu 2001)</span>.
Here we take advantage of the <code>samr</code> R package.</p>
<p>As reported from the <a href="https://tibshirani.su.domains/SAM/sam.pdf" class="external-link">SAM user guide</a>:</p>
<blockquote>
<p>The input to SAM is gene expression measurements from a set of
microarray experiments, as well as a response variable from each
experiment. The response variable may be a grouping like untreated,
treated (either unpaired or paired), a multiclass grouping (like breast
cancer, lymphoma, colon cancer), a quantitative variable (like blood
pressure) or a possibly censored survival time. SAM computes a statistic
d<sub>i</sub> for each gene <em>i</em>, measuring the strength of the
relationship between gene expression and the response variable. It uses
repeated permutations of the data to determine if the expression of any
genes are significantly related to the response. The cutoff for
significance is determined by a tuning parameter <strong>delta</strong>,
chosen by the user based on the false positive rate. One can also choose
a <strong>fold change</strong> parameter, to ensure that called genes
change at least a pre-specified amount.</p>
</blockquote>
<p>In <strong>featscreen</strong>, the SAM permutation test is computed
by calling the function <code><a href="../reference/rowSamStatistics.html">?rowSamStatistics</a></code>.</p>
<p>From the documentation, we can see that this is a wrapper function
that internally calls three functions from the <code>samr</code>
package:</p>
<ol style="list-style-type: decimal">
<li>
<code>samr</code>: correlate each feature with outcome
variable.</li>
<li>
<code>samr.compute.delta.table</code>: compute thresholds,
cutpoints, and false discovery rates for SAM analysis.</li>
<li>
<code>samr.compute.siggenes.table</code>: compute SAM statistics and
significance.</li>
</ol>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Define row/col size</span></span>
<span><span class="va">nr</span> <span class="op">=</span> <span class="fl">20</span></span>
<span><span class="va">nc</span> <span class="op">=</span> <span class="fl">20</span></span>
<span></span>
<span><span class="co">#Data</span></span>
<span><span class="va">x</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span>n <span class="op">=</span> <span class="va">nr</span><span class="op">*</span><span class="va">nc</span><span class="op">)</span>,</span>
<span>  nrow <span class="op">=</span> <span class="va">nr</span>,</span>
<span>  ncol <span class="op">=</span> <span class="va">nc</span>,</span>
<span>  dimnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"g"</span>,<span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="va">nr</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"S"</span>,<span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="va">nc</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#Categorical output vector (multinomial)</span></span>
<span><span class="va">y</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">3</span><span class="op">)</span>,size<span class="op">=</span><span class="va">nc</span>,replace<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"S"</span>,<span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="va">nc</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Compute</span></span>
<span><span class="fu"><a href="../reference/rowSamStatistics.html">rowSamStatistics</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x</span>,y<span class="op">=</span><span class="va">y</span><span class="op">)</span></span>
<span><span class="co">#&gt; $statistic</span></span>
<span><span class="co">#&gt;  [1]  0.421  0.026 -0.085  0.533 -0.766 -1.037  0.341 -0.169  0.966  0.322</span></span>
<span><span class="co">#&gt; [11]  0.115 -0.227  0.005 -0.021 -0.760 -0.087  0.533 -0.045 -0.344  0.406</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $significance</span></span>
<span><span class="co">#&gt;  [1] 0.8 0.8 0.8 0.8 0.8 0.8 0.8 0.8 0.8 0.8 0.8 0.8 0.8 0.8 0.8 0.8 0.8 0.8 0.8</span></span>
<span><span class="co">#&gt; [20] 0.8</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level2 unnumbered">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Tusher2001" class="csl-entry">
Tusher, Virginia Goss, Robert Tibshirani, and Gilbert Chu. 2001.
<span>“<span class="nocase">Significance analysis of microarrays applied
to the ionizing radiation response.</span>”</span> <em>Proceedings of
the National Academy of Sciences of the United States of America</em> 98
(9): 5116–21. <a href="https://doi.org/10.1073/pnas.091062498" class="external-link">https://doi.org/10.1073/pnas.091062498</a>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Alessandro Barberis.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
